processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -384,55 +381,115 @@ public int add(InputRow row) throws IndexSizeExceededException
       throw new IAE("Cannot add row[%s] because it is below the minTimestamp[%s]", row, new DateTime(minTimestamp));
     }
 
+    String[][] dims = null;
     final List<String> rowDimensions = row.getDimensions();
 
-    String[][] dims;
-    List<String[]> overflow = null;
-    synchronized (dimensionOrder) {
-      dims = new String[dimensionOrder.size()][];
-      for (String dimension : rowDimensions) {
-        List<String> dimensionValues = row.getDimension(dimension);
-
-        // Set column capabilities as data is coming in
-        ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
-        if (capabilities == null) {
-          capabilities = new ColumnCapabilitiesImpl();
-          capabilities.setType(ValueType.STRING);
-          columnCapabilities.put(dimension, capabilities);
+    for (String dimension : rowDimensions) {
+      List<String> dimensionValues = row.getDimension(dimension);
+
+      // Set column capabilities as data is coming in
+      ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
+      if (capabilities == null) {
+        capabilities = new ColumnCapabilitiesImpl();
+        capabilities.setType(ValueType.STRING);
+        ColumnCapabilitiesImpl other = columnCapabilities.putIfAbsent(dimension, capabilities);
+        if (null != other) {
```

Comment:
I can if that's the preferred style. I usually try to write null first in a boolean comparator as a habit from my C days, but I'll change it to be more consistent with the general Druid styling.